V1

mdaum opened the pull request:
first pass at npm watcher with tests...still have yet to add integration tests in which changes stream and/or producer fail...to make sure npmWatcher can recover...but this case is covered in unit tests w/ mocks

mdaum committed:
created skeleton for server, plans for actually hitting registry and stuff in README, added in configs and logging stuffs

mdaum committed:
addressed PR, using new queue util logic for connecting to rabbitmq

mdaum committed:
updated README with a more event-driven plan, I was misunderstanding how changes-stream actually works.

mdaum committed:
removing unnecessary timeouts since throttling will be enforced in checkWatcher

mdaum committed:
updated gitignore....setup loop done with logging and cleanup/restart logic

mdaum committed:
started fleshing out watch loop...and basic changesStream logic

mdaum committed:
killing superflous readme

mdaum committed:
scrapped original attempt...now using utils for producer creation...basic looping logic, right now running into error on creating the producer...

mdaum committed:
deleted unused file

mdaum committed:
now publishing to rabbitmq as changes come in...next up is some tests to verify that queue is acutally being loaded...and make sure error paths are solid

mdaum committed:
moved class into own js file for easier testing

mdaum committed:
unit tests for UpdatesPublisher...100% coverage

mdaum committed:
added stop method and test

mdaum committed:
added integration test for expected behaviour...will add error paths in another commit

mdaum committed:
getting rid of super huge test data I never used :S

mdaum committed:
moved some consts around in integration test file to make tests added in future a bit cleaner

ctdio commented on src/UpdatesPublisher.js:
```diff
@@ -0,0 +1,78 @@
+  }
+}
+
+module.exports = new UpdatesPublisher()
```
I don't think it's necessary to make the Updates publisher a singleton. It makes testing a little harder than it really needs to be (all tests now use the same instance of the publisher, forcing you to run everything serially). I think it would be better to just export the class itself.

mdaum replied:
Will rework it and run tests in parallel.

ctdio commented on test/integration/UpdatePublishersTest.js:
```diff
@@ -0,0 +1,48 @@
+  sandbox.restore()
+})
+
+test.serial('Will successfully detect changes and publish to rabbitmq', async (t) => {
```
As I mentioned above, moving away from a singleton `UpdatesPublisher` would allow future tests that directly use the publisher (non proxyquired) be run in parallel.

ctdio commented on test/unit/UpdatesPublisherTest.js:
```diff
@@ -0,0 +1,114 @@
+
+test.beforeEach('setup mock channel and connections', (t) => {
+  const producerOptions = {
+    queueName: 'test-queue'
```
add a `uuid` to this queueName

ctdio commented on test/unit/UpdatesPublisherTest.js:
```diff
@@ -0,0 +1,114 @@
+  sandbox.restore()
+})
+
+test.serial('#setupProducer', async (t) => {
```
If you make the change I recommended above (using a uuid in the test `queueName`, you should be able to run these tests in parallel since a new instance of `UpdatesPublisher` is create each time since you use the `noPreserveCache` option above.

ctdio commented on test/unit/UpdatesPublisherTest.js:
```diff
@@ -0,0 +1,114 @@
+test.afterEach('clean up', (t) => {
+  const { sandbox } = t.context
+
+  sandbox.restore()
```
Just to be safe, stop the publisher in this `afterEach`

mdaum replied:
stopping the publisher relies on it's changes and producer properties to be defined, not always a given as I am testing logic paths rather than the whole shebang...will leave it to the test to stop publisher when necessary

ctdio commented on src/UpdatesPublisher.js:
```diff
@@ -0,0 +1,78 @@
+    amqUrl = _amqUrl
+    registryUrl = _registryUrl
+    logger = _logger
+  }
```
When you move away from exporting this publisher as a singleton, you can make the `configure` method the class constructor. Just remember to attach the arguments to the object instead of having them floating at the top of the file.

ctdio commented on src/UpdatesPublisher.js:
```diff
@@ -0,0 +1,78 @@
+let registryUrl = null
+
+class UpdatesPublisher {
+  configure (_producerOptions, _amqUrl, _registryUrl, _logger) {
```
Quick tip: As a general rule of thumb, if your function takes in more than 2-3 arguments, you should consider making your function accept a single `options` object that contains the arguments you need. This ends up making code easier to read.

ex.
`definition`
```js
configure (options) {
  const { producerOptions, amqUrl, registryUrl, logger } = options
  // do work
}

// or
configure ({ producerOptions, amqUrl, registryUrl, logger }) {
  // do work 
}
```

`usage`
```js
publisher.configure({
  producerOptions: {},
  amqUrl: 'urlA', 
  registryUrl: 'urlB',
  logger: console
})

// vs
publisher.configure({}, 'urlA', 'urlB', console) 
// ^^^ easier to make mistakes and misplace args here
// also, I need to look at the definition to tell exactly what each arg is used for,
// it's much more obvious what is going on in the `options` arg version
```

mdaum replied:
Those cobrowse good habits tho

ctdio commented on test/unit/UpdatesPublisherTest.js:
```diff
@@ -0,0 +1,114 @@
+  const producerSendMessageSpy = sandbox.spy(publisher.producer, 'sendMessage')
+  publisher.changes.emitData()
+  await Promise.delay(100)
+  t.true(producerSendMessageSpy.calledWith({}))
```
The size of this test makes me wonder if there is a little too much being tested here. Maybe this can be split into multiple test cases? It would make debugging a failing test much easier.

mdaum replied:
can't shave off too much as the purpose of this test is to test the logic paths taken by the start method, and ensuring we hit what we want when our changes and producer fail...that being said I believe I can make two tests that explore the start method. One that will ensure we recover from a changes failure, and another that ensures we recover from a producer failure.

ctdio commented on test/unit/UpdatesPublisherTest.js:
```diff
@@ -0,0 +1,114 @@
+  await publisher.stop()
+  t.true(producerStopSpy.calledOnce)
+  t.true(changesDestroySpy.calledOnce)
+})
```
Some spacing in between major parts of these tests would help a lot with readability/maintainability. You had good comments in the test above this one, but the others are lacking some comments that would make them easier to follow.

ctdio replied:
Spacing makes things prettier and easier to parse.

ctdio commented on test/unit/UpdatesPublisherTest.js:
```diff
@@ -0,0 +1,114 @@
+  t.true(consoleSpy.calledWith('changes stream successfully created'))
+  publisher.changes.destroy()
+})
+test.serial('#start', async (t) => {
```
I like that you are specifying which function is being tested, but you should be adding more meaningful test descriptions. Exactly what is being tested here other than just the `start` function? What part of `start` is being tested? All of it? Also in what way is it significant? (see my comment below)

ctdio commented on test/unit/UpdatesPublisherTest.js:
```diff
@@ -0,0 +1,114 @@
+  }
+}
+
+test.beforeEach('setup mock channel and connections', (t) => {
```
+1 for a meaningful `beforeEach` description

mdaum committed:
changed UpdatesPublisher to export the whole class, not singleton...using options in constructor, tests now use queues with uuid's...need to make tests jive with changes to UpdatesPublisher and we should be good

mdaum committed:
adding package.lock....addressed PR comments related to converting from
singleton export to class export of UpdatesPublisher...modified test
script to run unit and integraiton in nyc ava runs for coverage purposes

mdaum committed:
renamed test...split out #start unit test...better descriptors...ready
for re-review

mdaum commented:
@charlieduong94 addressed your requested changes....new to github PR etiquette, there an official thing I do to notify you that I addressed your comments via the gui? Or do I just give a ping...

ctdio commented on package.json:
```diff
@@ -1,10 +1,14 @@
+    "docker:start": "docker-compose up -d",
+    "docker:stop": "docker-compose down",
+    "start": "node src/server.js",
+    "test": "npm run docker:start && sleep 10 && nyc ava test/unit/*.js && sleep 1 && nyc ava test/integration/*.js && sleep 1 && npm run docker:stop",
```
Can you change this to run all tests together? You'll just need to use `nyc ava test` instead and remove the `sleep 1`.

Also, add extra commands for just running unit tests and integration tests separately. 
```sh
# run unit tests only
npm run test:unit

# run integration tests only
npm run test:integration
```

ctdio commented on test/unit/UpdatesPublisherTest.js:
```diff
@@ -0,0 +1,152 @@
+  t.true(changesDestroySpy.notCalled)
+  await publisher.stop()
+  t.true(producerStopSpy.calledOnce)
+  t.true(changesDestroySpy.calledOnce)
```
For all of your spy assertions, you should be using methods exposed by `sandbox.assert`. The error messages are much better. If `ava` complains about no assertions being made, just add a `t.plan(0)` to the top of the test or alternatively a `t.pass()` at the bottom.

Epitrochoid commented:
@mdaum I usually add a comment like `Fixed.` if the change doesn't cause the original comment to be hidden. If the change hides the original comment that is generally enough to show that it was fixed.

ctdio commented on src/UpdatesPublisher.js:
```diff
@@ -0,0 +1,76 @@
+      logger.info('error emitted from producer, restarting it and changes stream')
+      this.changes.destroy()
+      await this.producer.stop()
+      await Promise.delay(1000)
```
Make these delay times constants at the top of the file. I don't like having arbitrary values floating around.

ctdio commented on test/mocks/mockProducer.js:
```diff
@@ -0,0 +1,20 @@
+module.exports = class mockProducer extends EventEmitter {
+  constructor () {
+    super()
+    this._stopped = false
```
never used?

ctdio commented on src/UpdatesPublisher.js:
```diff
@@ -0,0 +1,76 @@
+    this.producerOptions = options.producerOptions
+    this.amqUrl = options.amqUrl
+    this.registryUrl = options.registryUrl
+    this.logger = options.logger
```
For the fields that you append to your publisher, start them with an underscore to signify that they are supposed to be private and not really exposed to clients that are using the publisher.

```
constructor (options) {
  this._producerOptions = options.producerOptions
  this._amqUrl = options.amqUrl
  // etc..
}
```

ctdio commented on src/UpdatesPublisher.js:
```diff
@@ -0,0 +1,76 @@
+  async stop () {
+    const {logger} = this
+    this.changes.destroy()
+    await this.producer.stop()
```
I see that you have some weird checks in the tests that ensure that a producer and changes stream exists before you do your clean up. Just add the checks here instead.
```
if (this.changes) {
  this.changes.destroy()
}
// or 
this.changes && this.changes.destroy()
```

ctdio commented on src/UpdatesPublisher.js:
```diff
@@ -0,0 +1,76 @@
+      logger.info('producer successfully created')
+    } catch (error) {
+      logger.error(error)
+      logger.info('error creating producer, retrying ...')
```
This would be a bit better later on when we have some sort of logging system in place

```
logger.error('Error creating producer', error)
logger.info(`Attempting to reinitialize queue producer...`)
```

ctdio commented on src/UpdatesPublisher.js:
```diff
@@ -0,0 +1,76 @@
+    const {logger} = this
+    this.producer.on('error', async (error) => {
+      logger.error(error)
+      logger.info('error emitted from producer, restarting it and changes stream')
```
Similar to the above comment:

```
logger.error('Error received from producer', error)
logger.info('Restarting producer and changes stream...')
```

ctdio commented on src/config/Config.js:
```diff
@@ -0,0 +1,24 @@
+    },
+
+    registryUrl: {
+      description: 'URL for watcher\'s registry',
```
`URL for npm registry`

ctdio reviewed (changes requested):
Nice changes @mdaum. Just a few more minor things that need to be changed.

btw, there's a `.DS_Store` file that was added, you should remove that and add rule to your `.gitignore` to ensure that those never get added again.

ctdio commented:
Oh and btw, Github sends notifications when you make updates or comments to PRs that you are a reviewer on. But if you want me to review something quickly or if you feel like you've been waiting too long, feel free to ping me on slack.

mdaum committed:
addressed PR comments

ctdio reviewed (approved):
@mdaum nice job with this.
